streams/src/main/java/org/apache/kafka/streams/state/internals/LogicalKeyValueSegment.java

```diff
@@ -0,0 +1,297 @@
+            prefixKeyFormatter.forPhysicalStore(from),
+            prefixKeyFormatter.forPhysicalStore(to),
+            openIterators);
+        return new StrippedPrefixKeyValueIteratorAdapter(
+            iteratorWithKeyPrefixes,
+            prefixKeyFormatter::fromPhysicalStore);
+    }
+
+    @Override
+    public synchronized KeyValueIterator<Bytes, byte[]> all() {
+        final KeyValueIterator<Bytes, byte[]> iteratorWithKeyPrefixes = physicalStore.prefixScan(
+            prefixKeyFormatter.getPrefix(),
+            new BytesSerializer(),
+            openIterators);
+        return new StrippedPrefixKeyValueIteratorAdapter(
+            iteratorWithKeyPrefixes,
+            prefixKeyFormatter::fromPhysicalStore);
+    }
+
+    @Override
+    public long approximateNumEntries() {
+        throw new UnsupportedOperationException("cannot estimate num entries for logical segment");
+    }
+
+    @Override
+    public void addToBatch(final KeyValue<byte[], byte[]> record, final WriteBatch batch) throws RocksDBException {
+        physicalStore.addToBatch(
+            new KeyValue<>(
+                prefixKeyFormatter.forPhysicalStore(record.key),
+                record.value),
+            batch);
+    }
+
+    @Override
+    public void write(final WriteBatch batch) throws RocksDBException {
+        // no key transformations here since they should've already been done as part
+        // of adding to the write batch
+        physicalStore.write(batch);
+    }
+
+    /**
+     * Manages translation between raw key and the key to be stored into the physical store.
+     * The key for the physical store is the raw key prepended with a fixed-length prefix.
+     */
+    private static class PrefixKeyFormatter {
+        private final byte[] prefix;
+
+        PrefixKeyFormatter(final Bytes prefix) {
+            this.prefix = prefix.get();
+        }
+
+        Bytes forPhysicalStore(final Bytes key) {
+            return key == null ? null : Bytes.wrap(forPhysicalStore(key.get()));
+        }
+
+        byte[] forPhysicalStore(final byte[] key) {
+            final byte[] keyWithPrefix = new byte[prefix.length + key.length];
+            System.arraycopy(prefix, 0, keyWithPrefix, 0, prefix.length);
+            System.arraycopy(key, 0, keyWithPrefix, prefix.length, key.length);
+            return keyWithPrefix;
+        }
+
+        Bytes fromPhysicalStore(final Bytes keyWithPrefix) {
+            return Bytes.wrap(fromPhysicalStore(keyWithPrefix.get()));
+        }
+
+        private byte[] fromPhysicalStore(final byte[] keyWithPrefix) {
+            final int rawKeyLength = keyWithPrefix.length - prefix.length;
+            final byte[] rawKey = new byte[rawKeyLength];
+            System.arraycopy(keyWithPrefix, prefix.length, rawKey, 0, rawKeyLength);
+            return rawKey;
+        }
+
+        Bytes getPrefix() {
+            return Bytes.wrap(prefix);
```

Comment:
`wrap()` just creates the new object (after performing a null check) so it's very lightweight. 

It's more convenient to keep `prefix` as `byte[]` than `Bytes` because all the other operations require `byte[]` rather than `Bytes`. If we really wanted we could keep both (one copy as `byte[]` and another as `Bytes`) but that feels like overkill.